src/postprocess/chi0.jl

```diff
@@ -134,34 +132,137 @@ LinearAlgebra.ldiv!(P::FunctionPreconditioner, x) = (x .= P.precondition!(simila
     δψnk
 end
 
-"""
-Returns the change in density `δρ` for a given `δV`. Drop all non-diagonal terms with
-(f(εn)-f(εm))/(εn-εm) factor less than `droptol`. If `sternheimer_contribution`
-is false, only compute excitations inside the provided orbitals.
+# Returns δψ, δocc, δεF corresponding to a change in *total* Hamiltonian dH
+# P = f_fd(H-εF), tr(P) = N
+# δεn = <ψn|δV|ψn>
+# 0 = ∑_n f'n (δεn - δεF) determines δεF
 
-Note: This function assumes that all bands contained in `ψ` and `eigenvalues` are
-sufficiently converged. By default the `self_consistent_field` routine of `DFTK`
-returns `3` extra bands, which are not converged by the eigensolver
-(see `n_ep_extra` parameter). These should be discarded before using this function.
-"""
-@views @timing function apply_χ0(ham, ψ, εF, eigenvalues, δV;
-                          droptol=0,
-                          sternheimer_contribution=true,
-                          temperature=ham.basis.model.temperature,
-                          kwargs_sternheimer=(cgtol=1e-6, verbose=false))
+# Then <ψm|δP|ψn> = (fm-fn)/(εm-εn) <ψm|δH|ψn>
+# Except for the diagonal which is
+# <ψn|δP|ψn> = (fn'-δεF) <ψn|δH|ψn>
+
+# We want to represent this with a δψ, δf, we do *not* impose that
+# δψ is orthogonal at finite temp.
+# We get
+# δP = ∑_k fk |δψk><ψk| + |ψk><δψk| + δ fk |ψk><ψk|
+# Identifying with <ψm|δP|ψn> we get for the diagonal terms
+# δfn = fn'<ψn|δH-δεF|ψn>
+# and for the off-diagonal
+# (fm-fn)/(εm-εn) <ψm|δH|ψn> = fm <δψm|ψn> + fn <ψm|δψn>
+# We split the computation of δψn in two contributions:
+# for the already-computed states, we add an explicit contribution
+# for the empty states, we solve a Sternheimer equation
+# (H-εn) δψn = - P_{ψ^⟂} δH ψn
+
+# The explicit term needs a careful consideration of stability.
+# Let <ψm|δψn> = αmn <δψm|δH|ψn>. αmn has to satisfy
+# fn αmn + fm αnm = ratio = (fn-fm)/(εn-εm)
+# The usual way is to impose orthogonality (=> αmn=-αnm),
+# but this means that αmn = 1/(εm-εn), which is unstable
+# Instead, we minimize αmn^2 + αnm^2 under the linear constraint above, which leads to
+# fn αmn = ratio * fn^2 / (fn^2 + fm^2)
+
+# This formula is very nice
+# - It gives a vanishing contribution fn αmn for empty states (note that α itself blows up, but it's compensated by fn)
+# - In the case where fn=1/0 or fm=0 we recover the same formulas as the ones with orthogonality
+# - It does not blow up for degenerate states
+function get_αmn(fm, fn, ratio)
+    ratio == 0 && return ratio
+    ratio * fn / (fn^2 + fm^2)
+end
+
+@views function solve_Ω(ham, ψ, occ, εF, eigenvalues, δHψ; kwargs_sternheimer=(cgtol=1e-6, verbose=false))
```

Comment:
I remember you telling me that you use `occupation` and not `occ` :stuck_out_tongue: